subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -63,270 +40,265 @@ static bool is_entry_defined(struct bt_mesh_scheduler_srv *srv, uint8_t idx)
-		31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
-	struct tm revised_time = *current_local;
-	bool is_year_revised = false;
+	int day_cnt = 0;
 
-	revised_time.tm_mday += number_days;
+	year += TM_START_YEAR;
 
-	if (revised_time.tm_mday > days[revised_time.tm_mon]) {
-		revised_time.tm_mday -= days[revised_time.tm_mon];
-		if (revised_time.tm_mon == DECEMBER) {
-			revised_time.tm_mon = JANUARY;
-			revised_time.tm_year++;
-			is_year_revised = true;
-		} else {
-			revised_time.tm_mon++;
-		}
+	for (int i = TM_START_YEAR; i < year; i++) {
+		day_cnt += is_leap_year(i) ? DAYS_LEAP_YEAR : DAYS_YEAR;
 	}
 
-	if (is_year_revised) {
-		if (!revise_year(sched_time, current_local, entry, 1)) {
-			return false;
-		}
-	}
+	int days[12] = {31,
+		is_leap_year(year) ? 29 : 28,
+		31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
 
-	sched_time->tm_mday = revised_time.tm_mday;
-	if (revised_time.tm_mon != current_local->tm_mon) {
-		return set_month(sched_time, &revised_time, entry);
+	for (int i = 0; i < month; i++) {
+		day_cnt += days[i];
 	}
 
-	return true;
+	day_cnt += day;
+	return (day_cnt - 1) % WEEKDAY_CNT;
 }
 
-static bool revise_day(struct tm *sched_time,
-		       struct tm *current_local,
-		       struct bt_mesh_schedule_entry *entry,
-		       struct bt_mesh_time_srv *srv,
-		       int number_days)
+static bool day_validation(struct tm *sched_time, int day)
 {
-	struct tm revised_time = *current_local;
+	int days[12] = {31,
+		is_leap_year(sched_time->tm_year + TM_START_YEAR) ? 29 : 28,
+		31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
 
-	revised_time.tm_mday += number_days;
-	return set_day(sched_time, &revised_time, entry, srv);
+	return day <= days[sched_time->tm_mon];
 }
 
-static bool revise_hour(struct tm *sched_time,
-			struct tm *current_local,
-			struct bt_mesh_schedule_entry *entry,
-			struct bt_mesh_time_srv *srv,
-			int number_hours)
+static bool convert_scheduler_time_to_tm(struct tm *sched_time,
+					struct tm *current_local,
+					struct bt_mesh_schedule_entry *entry)
 {
-	struct tm revised_time = *current_local;
+	bool consider_ovflw;
+	bool second_ovflw;
+	bool minute_ovflw;
+	bool hour_ovflw;
+	bool day_ovflw;
+	int month;
 
-	revised_time.tm_hour += number_hours;
-	return set_hour(sched_time, &revised_time, entry, srv);
-}
+	if (entry->month == 0) {
+		return false;
+	}
 
-static bool revise_minute(struct tm *sched_time,
-			  struct tm *current_local,
-			  struct bt_mesh_schedule_entry *entry,
-			  struct bt_mesh_time_srv *srv,
-			  int number_minutes)
-{
-	struct tm revised_time = *current_local;
+	if (entry->day_of_week == 0) {
+		return false;
+	}
 
-	revised_time.tm_min += number_minutes;
-	return set_minute(sched_time, &revised_time, entry, srv);
-}
+	int start_year = current_local->tm_year;
 
-static bool set_year(struct tm *sched_time,
-		     struct tm *current_local,
-		     struct bt_mesh_schedule_entry *entry)
-{
-	uint8_t current_year = current_local->tm_year % 100;
+year_set:
```

Comment:
Implementation requires cycling since special action configuration values cause time revision. 
The previous version caused multiple implicit recursion that was very difficult to debug.
The current version is possible to change by using the `while` loop and a subset of flags to highlight which stage to revise. However, it will work slower. 
I do not share the ostracism that `goto` is subjected to. If it is possible to use `goto` to make the algorithm simpler it is worth to use.

Reply:
This is fine by me. But as per my understanding, nRF Connect SDK seems to follow Zephyr Coding Guidelines, which follow Misra C requirements, which prevents this specific usage of `goto`. Also, I'm not sure if static analyzer will not complain about this.

Reply:
@trond-snekvik , should it be reworked to get rid of `goto up`?